### PR TITLE
Add support for logprobs in OpenAI chat API

### LINKED ISSUE #839: [Bug] Chat completions logprobs support
### Checklist

- [X] 1. I have searched related issues but cannot get the expected help.
- [X] 2. The bug has not been fixed in the latest version.
- [X] 3. Please note that if the bug-related issue you submitted lacks corresponding environment info and a minimal reproducible demo, it will be challenging for us to reproduce and resolve the issue, reducing the likelihood of receiving feedback.

### Describe the bug

I'm encountering an issue when making a request to the `v1/chat/completions` endpoint with the `"logprobs": true` parameter. The `choices.logprobs` field in the response is always null, even though it should be populated with probability values.

### Reproduction

Script to launch the SGLang server:

```bash
python -m sglang.launch_server  \
    --model-path Qwen/Qwen2-7B-Instruct \
    --port 8000
```

Script to reproduce the request:

```bash
curl http://127.0.0.1:8000/v1/chat/completions \
  -H "Content-Type: application/json" \
  -d '{
     "model": "Qwen/Qwen2-7B-Instruct",
     "messages": [{"role": "user", "content": "Hi!"}],
     "stream": false,
     "temperature": 1,
     "max_tokens": 1,
     "top_logprobs": 2,
     "logprobs": true
   }'
```

Response:

```json
{"id":"704abb89338b4892bd63b877764f9593","object":"chat.completion","created":1722369690,"model":"Qwen/Qwen2-7B-Instruct","choices":[{"index":0,"message":{"role":"assistant","content":"Hello"},"logprobs":null,"finish_reason":"FINISH_LENGTH: 1"}],"usage":{"prompt_tokens":21,"total_tokens":22,"completion_tokens":1}}
```

### Environment

```Shell
Python: 3.10.14 (main, May  6 2024, 19:42:50) [GCC 11.2.0]
CUDA available: True
GPU 0: NVIDIA A100-SXM4-80GB
CUDA_HOME: /usr/local/cuda
NVCC: Cuda compilation tools, release 12.3, V12.3.107
CUDA Driver Version: 525.105.17
PyTorch: 2.3.1
sglang: 0.2.7
flashinfer: 0.1.2+cu121torch2.3
requests: 2.32.3
tqdm: 4.66.4
numpy: 1.26.4
aiohttp: 3.9.5
fastapi: 0.111.1
hf_transfer: 0.1.8
huggingface_hub: 0.24.3
interegular: 0.3.3
packaging: 24.1
PIL: 10.4.0
psutil: 6.0.0
pydantic: 2.8.2
uvicorn: 0.30.3
uvloop: 0.19.0
zmq: 26.0.3
vllm: 0.5.3.post1
openai: 1.37.1
anthropic: 0.32.0
NVIDIA Topology: 
        GPU0    NIC0    NIC1    NIC2    NIC3    NIC4    NIC5    NIC6    NIC7    NIC8    NIC9    CPU Affinity    NUMA Affinity
GPU0     X      SYS     SYS     SYS     SYS     SYS     SYS     NODE    NODE    PXB     PXB     96-127,224-255  3
NIC0    SYS      X      PIX     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS
NIC1    SYS     PIX      X      SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS
NIC2    SYS     SYS     SYS      X      PIX     SYS     SYS     SYS     SYS     SYS     SYS
NIC3    SYS     SYS     SYS     PIX      X      SYS     SYS     SYS     SYS     SYS     SYS
NIC4    SYS     SYS     SYS     SYS     SYS      X      PXB     SYS     SYS     SYS     SYS
NIC5    SYS     SYS     SYS     SYS     SYS     PXB      X      SYS     SYS     SYS     SYS
NIC6    NODE    SYS     SYS     SYS     SYS     SYS     SYS      X      PIX     NODE    NODE
NIC7    NODE    SYS     SYS     SYS     SYS     SYS     SYS     PIX      X      NODE    NODE
NIC8    PXB     SYS     SYS     SYS     SYS     SYS     SYS     NODE    NODE     X      PXB
NIC9    PXB     SYS     SYS     SYS     SYS     SYS     SYS     NODE    NODE    PXB      X 

Legend:

  X    = Self
  SYS  = Connection traversing PCIe as well as the SMP interconnect between NUMA nodes (e.g., QPI/UPI)
  NODE = Connection traversing PCIe as well as the interconnect between PCIe Host Bridges within a NUMA node
  PHB  = Connection traversing PCIe as well as a PCIe Host Bridge (typically the CPU)
  PXB  = Connection traversing multiple PCIe bridges (without traversing the PCIe Host Bridge)
  PIX  = Connection traversing at most a single PCIe bridge
  NV#  = Connection traversing a bonded set of # NVLinks

NIC Legend:

  NIC0: mlx5_0
  NIC1: mlx5_1
  NIC2: mlx5_2
  NIC3: mlx5_3
  NIC4: mlx5_4
  NIC5: mlx5_5
  NIC6: mlx5_6
  NIC7: mlx5_7
  NIC8: mlx5_8
  NIC9: mlx5_9


ulimit soft: 1048576
```
```

### PR DESCRIPTION
**Pull request by yichuan-w**
Thank you for your contribution, we really appreciate it. The following instructions will help improve your pull request and make it easier to receive feedback. If there are any items you don't understand, don't worry. Just submit the pull request and ask the maintainers for help.

## Motivation

Fix #839 

## Modification

Add support for logprobs in OpenAI chat API and cater real OAI API output

## Checklist

1. Ensure pre-commit `pre-commit run --all-files` or other linting tools are used to fix potential lint issues.
2. Confirm that modifications are covered by complete unit tests. If not, please add more unit tests for correctness.
3. Modify documentation as needed, such as docstrings or example tutorials.

### CODE REVIEW

**Comment by yichuan-w** [comment]
cc @merrymercy @hnyls2002 @Ying1123 @zhyncs for review, I think it is ready to merge

File: python/sglang/srt/openai_api/protocol.py [hunk 1]
```diff
@@ -54,6 +54,24 @@ class LogProbs(BaseModel):
...
+
+class ChatCompletionTokenLogprob(BaseModel):
+    token: str
+    bytes: List[int]
+    logprob: float
+    top_logprobs: List[TopLogprob]
+
+
+class ChoiceLogprobs(BaseModel):
+    ## build for v1/chat/completions response
```
Comment (merrymercy) [inline python/sglang/srt/openai_api/protocol.py:71]: why do you use "##" for comment instead of "#"

Comment (merrymercy) [inline python/sglang/srt/openai_api/protocol.py]: why do you use "##" instead of "#"?

**Review by merrymercy — APPROVED** [review]

**Comment by isukharev** [comment]
I've just tested the code from this branch and encountered an error:

```
Exception in ControllerSingle:
Traceback (most recent call last):
  File "/home/user/sources/sglang/python/sglang/srt/managers/controller_single.py", line 166, in start_controller_process
    controller.loop_for_forward()
  File "/home/user/sources/sglang/python/sglang/srt/managers/controller_single.py", line 103, in loop_for_forward
    out_pyobjs = self.tp_server.exposed_step(recv_reqs)
  File "/home/user/sources/sglang/python/sglang/srt/managers/tp_worker.py", line 209, in exposed_step
    self.forward_step()
  File "/home/user/sglang/lib/python3.10/site-packages/torch/utils/_contextlib.py", line 115, in decorate_context
    return func(*args, **kwargs)
  File "/home/user/sources/sglang/python/sglang/srt/managers/tp_worker.py", line 225, in forward_step
    self.forward_prefill_batch(new_batch)
  File "/home/user/sources/sglang/python/sglang/srt/managers/tp_worker.py", line 532, in forward_prefill_batch
    output = self.model_runner.forward(batch, ForwardMode.EXTEND)
  File "/home/user/sources/sglang/python/sglang/srt/model_executor/model_runner.py", line 386, in forward
    return self.forward_extend(batch)
  File "/home/user/sglang/lib/python3.10/site-packages/torch/utils/_contextlib.py", line 115, in decorate_context
    return func(*args, **kwargs)
  File "/home/user/sources/sglang/python/sglang/srt/model_executor/model_runner.py", line 354, in forward_extend
    return self.model.forward(
  File "/home/user/sglang/lib/python3.10/site-packages/torch/utils/_contextlib.py", line 115, in decorate_context
    return func(*args, **kwargs)
  File "/home/user/sources/sglang/python/sglang/srt/models/qwen2.py", line 288, in forward
    return self.logits_processor(
  File "/home/user/sglang/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1532, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/home/user/sglang/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1541, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/user/sources/sglang/python/sglang/srt/layers/logits_processor.py", line 209, in forward
    all_logits = all_logits[:, : self.config.vocab_size].float()
torch.cuda.OutOfMemoryError: CUDA out of memory. Tried to allocate 8.71 GiB. GPU
```

**Comment by hnyls2002** [comment]
@isukharev Computing logprobs take more memory, try to reduce `--mem-fraction-static` and we are figuring out how to trade-off between the memory usage and bigger batch size.

**Comment by isukharev** [comment]
@hnyls2002 Thanks! After reducing `--mem-fraction-static` to `0.7`, the previous error gone, but the generation speed relative to the current master version dropped by a factor of 15, and now the cache hit rate is consistently 0.00% instead of the previous cache hit rate around 90%.

File: python/sglang/srt/openai_api/protocol.py (hunk 1 above)
Comment (yichuan-w) [reply python/sglang/srt/openai_api/protocol.py]: fixed

**Comment by yichuan520030910320** [comment]
I think it is ready to merge now

**Comment by Ying1123** [comment]
@isukharev  When you want to get the logprob of the prompt, the prefix radix cache will be turned off. This is because we only cache the KV cache, not the logits.

Do you need logprob for prompts or do you only need the logprob for generation? If you only need the logprob for generation, we can implement an additional flag just for your use case. Then you should see a similar cache hit rate.

**Comment by isukharev** [comment]
@Ying1123 Only need this for generation, we are using LLM as a CrossEncoder as shown here: 
https://github.com/openai/openai-cookbook/blob/main/examples/Search_reranking_with_cross-encoders.ipynb

**Comment by Ying1123** [comment]
@yichuan520030910320 @hnyls2002 It should be easy to do. Can you implement this feature?

**Pull request merged**